### DISCUSSION TITLE
False positive on tiktok videos

### DISCUSSION THREAD

**Discussion by azurenekowo** [infinitered/nsfwjs#705] [Model Training]
I had a discord bot up and running this module, apparently [this image](https://media.discordapp.net/attachments/899255260546203659/924311976165912596/unknown.png) is flagging for "Porn" with 62%???

I was running Inception v3 (the default model). 
![image](https://user-images.githubusercontent.com/67498765/147388501-8e7a3ca6-814e-480f-a7ac-0385e3a21b5f.png)

**Comment by azurenekowo** [comment]
Update: Somehow mangas / doujinshis got flagged aswell 

[This image](https://media.discordapp.net/attachments/899255260546203659/924507692108759120/FB_IMG_1640490167151.jpg) flagged with 98% hentai...

**Comment by GantMan** [comment]
There will be false positives for sure.  Can you provide the images in this chat (that link isn't working) and also can you identify which model you were using?

**Comment by azurenekowo** [comment]
@GantMan sure, here is the image

![Screenshot_20211228-230823_Gallery](https://user-images.githubusercontent.com/67498765/147584973-885286a7-de79-4748-b258-eb2f8b8b6d64.jpg)

I am using the default model (Inception v3) as the same as the example in the readme file.

**Comment by azurenekowo** [comment]
UPDATE: IT IS NOT INCEPTION V3. I just do nsfw.load() without any arguments. 

![Screenshot_20211228-231159_Chrome](https://user-images.githubusercontent.com/67498765/147585359-1cf349bb-2919-4576-a031-9250ee894d66.jpg)

**Comment by GantMan** [comment]
I just tested this on all three models.  All 3 said top class was Neutral.   While the one you're showing there is barely sure, it's still correct.  Which one is saying 62%?

**Comment by azurenekowo** [comment]
@GantMan hello, I don't know what module I'm currently using, I just copied the example nodejs code in the readme file.

![image](https://user-images.githubusercontent.com/67498765/147620122-d6ecc324-206f-41e0-b990-4b01c24d4a87.png)

By the way, I found the original image. The image I sent 7 hours ago was a screenshot.

![image](https://user-images.githubusercontent.com/67498765/147620142-f5bcf2df-5835-45ff-a9d1-495217b8ceb0.png)

**Comment by azurenekowo** [comment]
Also, when I upload the bot to heroku, it seems like calling image.dispose() will cause a memory leak. Any solution on this?

![image](https://user-images.githubusercontent.com/67498765/147620345-0bbbc6cb-0791-4ddb-85f7-a0d9344e80b7.png)

![image](https://user-images.githubusercontent.com/67498765/147620339-c5beb83b-994f-42af-89cd-b644b3518b1b.png)

**Comment by azurenekowo** [comment]
More updates: Mangas also got false-flagged.

![image](https://user-images.githubusercontent.com/67498765/147622071-49bf4b97-d465-473c-9d5c-ac41c8cefc27.png)

![image](https://user-images.githubusercontent.com/67498765/147622098-cd34f501-b398-4989-874c-5481f02f96c2.png)

**Comment by GantMan** [comment]
Hrmmm I'm not sure what the issue is right away.   Strange that the result is different from what you see on NSFWJS.com

Also, I'm not sure where your memory leak is coming from either.

I'm a bit busy, so I can't really spin up a node instance to debug at this time.  Can you check the image Tensor you're getting against this one?

Here, in this code, I grab the first 100 values of your image: https://bit.ly/3EBYYSi
Can you grab the first 100 values of your tensor image and compare?

**Comment by azurenekowo** [comment]
> Hrmmm I'm not sure what the issue is right away. Strange that the result is different from what you see on NSFWJS.com
> 
> Also, I'm not sure where your memory leak is coming from either.
> 
> I'm a bit busy, so I can't really spin up a node instance to debug at this time. Can you check the image Tensor you're getting against this one?
> 
> Here, in this code, I grab the first 100 values of your image: https://bit.ly/3EBYYSi Can you grab the first 100 values of your tensor image and compare?

Hello, sorry for the late reply sir, here is the result.

![image](https://user-images.githubusercontent.com/67498765/147718536-639b7f35-36cd-4209-841b-b1635958458a.png)

The code that I used:

![image](https://user-images.githubusercontent.com/67498765/147718555-9eba2d40-6af0-402b-bfb7-34a7a0d99a80.png)

**Comment by GantMan** [comment]
Well, the first thing I notice is that the tensors have different values.  They don't seem terribly off from one another, but they definitely have different values.   That's probably the main issue.   I'll see if I can figure out why, you should too.

![image](https://user-images.githubusercontent.com/997157/147778986-a34d5256-40c3-433c-aef7-a020bafb8ad4.png)

**Comment by azurenekowo** [comment]
New false postivite: [Rule34 link](https://wimg.rule34.xxx//images/3321/6d4255f47c171338b5b05ee55a171df89b64b84e.png)

Tested by all 3 modules at https://nsfwjs.com/
All 3 returned "Drawing" for > 97%

@GantMan

**Comment by GantMan** [comment]
That link is what we call a "false negative".  I'll add it to the dataset for future training.
